Firmware/motion_control.cpp

```diff
@@ -25,121 +26,140 @@
-     vector rotations. This requires only two cos() and sin() computations to form the rotation
-     matrix for the duration of the entire arc. Error may accumulate from numerical round-off, since
-     all double numbers are single precision on the Arduino. (True double precision will not have
-     round off issues for CNC applications.) Single precision error can accumulate to be greater than
-     tool precision in some cases. Therefore, arc path correction is implemented. 
-
-     Small angle approximation may be used to reduce computation overhead further. This approximation
-     holds for everything, but very small circles and large mm_per_arc_segment values. In other words,
-     theta_per_segment would need to be greater than 0.1 rad and N_ARC_CORRECTION would need to be large
-     to cause an appreciable drift error. N_ARC_CORRECTION~=25 is more than small enough to correct for 
-     numerical drift error. N_ARC_CORRECTION may be on the order a hundred(s) before error becomes an
-     issue for CNC machines with the single precision Arduino calculations.
-     
-     This approximation also allows mc_arc to immediately insert a line segment into the planner 
-     without the initial overhead of computing cos() or sin(). By the time the arc needs to be applied
-     a correction, the planner should have caught up to the lag caused by the initial mc_arc overhead. 
-     This is important when there are successive arc motions. 
-  */
-  // Vector rotation matrix values
-  float cos_T = 1-0.5*theta_per_segment*theta_per_segment; // Small angle approximation
-  float sin_T = theta_per_segment;
-  
-  float arc_target[4];
-  float sin_Ti;
-  float cos_Ti;
-  float r_axisi;
-  uint16_t i;
-  int8_t count = 0;
-
-  // Initialize the linear axis
-  arc_target[axis_linear] = position[axis_linear];
-  
-  // Initialize the extruder axis
-  arc_target[E_AXIS] = position[E_AXIS];
-
-  for (i = 1; i<segments; i++) { // Increment (segments-1)
-    
-    if (count < N_ARC_CORRECTION) {
-      // Apply vector rotation matrix 
-      r_axisi = r_axis0*sin_T + r_axis1*cos_T;
-      r_axis0 = r_axis0*cos_T - r_axis1*sin_T;
-      r_axis1 = r_axisi;
-      count++;
-    } else {
-      // Arc correction to radius vector. Computed only every N_ARC_CORRECTION increments.
-      // Compute exact location by applying transformation matrix from initial radius vector(=-offset).
-      cos_Ti = cos(i*theta_per_segment);
-      sin_Ti = sin(i*theta_per_segment);
-      r_axis0 = -offset[axis_0]*cos_Ti + offset[axis_1]*sin_Ti;
-      r_axis1 = -offset[axis_0]*sin_Ti - offset[axis_1]*cos_Ti;
-      count = 0;
+void mc_arc(float* position, float* target, float* offset, float feed_rate, float radius, uint8_t isclockwise, uint8_t extruder)
+{
+    // Extract the position to reduce indexing at the cost of a few bytes of mem
+    float p_x = position[X_AXIS];
+    float p_y = position[Y_AXIS];
+    float p_z = position[Z_AXIS];
+    float p_e = position[E_AXIS];
```

Comment:
We are quite tight on SRAM. Is it that much of a performance hit if you use `position` and `target` directly in the code? I'm pretty sure the compiler has some optimization enabled for indexing with a const value.

Reply:
I can switch them back quite easily.  It was noticeable, but not huge.  As an added benefit, clamp to endstops could be added for each segment.  Do you think I should add that too (see the commented out code right before the call to plan_buffer_line) after I convert back to using target and position arrays?

Reply:
It would be best if we clamped to software endstops.  Btw, how well does Crash Detection work during arcs?

Reply:
Alas, I only have an MK2.5, so I'm not sure (I don't think this feature is supported until MK3, but I could be wrong).  It would be particularly difficult to recover in the middle of an arc.  I believe we would have to save the arc interpolation settings and the segment number that failed in order to guarantee printing starts at the exact same place.  It is definitely doable though, at least if I understand Crash Detection properly (which I may not).

Reply:
The save and resume code is pretty complex. I’m quite sure that it is not handled correctly with G2/3 (and it never was, so it’s not your fault).